draft-ietf-httpbis-http2bis.xml

```diff
@@ -2934,16 +2934,29 @@ cookie: e=f
                 pseudo-header field to convey authority information, unless there is no authority
                 information to convey (in which case it MUST NOT generate :authority).
               </t>
+              <t>
+                Clients MUST NOT generate a request with a <tt>Host</tt> header field that differs
+                from the <tt>:authority</tt> pseudo-header field.  A server MAY treat a request as
+                malformed if it contains a <tt>Host</tt> header field that is different from the
```

Comment:
What is 'different' here? If one is present and one isn't, is that different? What if they have case differences? Etc.

Reply:
I didn't specify that deliberately, allowing someone to treat "example.com" and "example.com:443" and "Example.com" each as different to the other.

Would it be best to point that out?

> Servers are not required to normalize these fields before comparing them, which means that clients that produce both fields need to make the values byte-for-byte identical.

Reply:
or just add 'when their values are compared, byte-for-byte'

Reply:
They should not be compared byte-for-byte, we did this mistake recently and broke a few clients :-)
RFC3986#6.2.3 explains the scheme-based normalization, that works particularly well for H2 (in short, drop ":80" on "http" and ":443" on "https").

I think I'm fine with a SHOULD though. As Stefan (I think) mentioned, most implementations currently do not perform the comparison and a MUST would instantly make them non-compliant.

HTTP/1.1 messaging proceeds differently (in 3.2). It enumerates what a client must send, then says that a recipient MUST reject as badreq any request with invalid or missing Host. This leaves a bit of gray area about what you consider as invalid but does mandate that *some* checks are performed. I'm not sure I like this approach better, but I wanted to note it if that can fuel the discussion.

Reply:
I definitely think we should leave a breadcrumb to, say, scheme-based normalization, as well as to the relevant bit of the -semantics draft ([§ 7.2 Host and Authority](https://www.ietf.org/archive/id/draft-ietf-httpbis-semantics-19.html#name-host-and-authority)).

Reply:
The problem with citing RFC 3986 is that it doesn't actually specify anything in Section 6.2.  It's all examples.

Section 7.2 of -semantics is similarly unhelpful.

However, as the server is authoritative, perhaps we can rely on its own definitions of authority and say if the fields identify different authorities - by its own definition - it can reject the request.  Would that work?

Reply:
I guess we'd apply the same guidance to intermediaries? I think that's probably acceptable from where I'm sitting.

Reply:
OK, I've worked something out.  It's pretty gross though.  Let me know what you think.